golang1.15 build error : misplaced compiler directive

Hi, I just got this error when I'm running CI with golang1.15 :
```
/go/pkg/mod/github.com/!gitbook!i!o/syncgroup@v0.0.0-20181003125046-3e73b2e6a972/quickhash/strhash.go:8:3: misplaced compiler directive
```
seems this kind of grammar is not allowed on golang1.15
```golang
//go:noescape
//go:linkname useAeshash runtime.useAeshash
var useAeshash bool
```

--- a/go.mod
+++ b/go.mod
@@ -1,0 +1,3 @@
+module github.com/GitbookIO/syncgroup
+
+go 1.15

--- a/mutex/runtime.go
+++ b/mutex/runtime.go
@@ -4,7 +4,6 @@ import (
 	"unsafe"
 )
 
-//go:noescape
 //go:linkname race_Enabled internal.race.Enabled
 var race_Enabled bool
 

--- a/quickhash/strhash.go
+++ b/quickhash/strhash.go
@@ -1,40 +1,40 @@
 package quickhash
 
 import (
+	"reflect"
 	_ "runtime"
 	"unsafe" // required to use //go:linkname
 )
 
-//go:noescape
-//go:linkname useAeshash runtime.useAeshash
-var useAeshash bool
+// Fixed seed for deterministic outcomes
+const shSeed uint64 = 42
 
-//go:noescape
-//go:linkname strhash runtime.strhash
-func strhash(a unsafe.Pointer, h uintptr) uintptr
-
-//go:noescape
-//go:linkname aeshashstr runtime.aeshashstr
-func aeshashstr(p unsafe.Pointer, h uintptr) uintptr
-
-func StrHash(str string) uint64 {
-	if useAeshash {
-		return uint64(aeshashstr(unsafe.Pointer(&str), 0))
+// Modified from runtime/alg.go
+func rthash(ptr unsafe.Pointer, size int, seed uint64) uint64 {
+	if size == 0 {
+		return seed
 	}
-	return uint64(strhash(unsafe.Pointer(&str), 0))
+	// The runtime hasher only works on uintptr. For 64-bit
+	// architectures, we use the hasher directly. Otherwise,
+	// we use two parallel hashers on the lower and upper 32 bits.
+	if unsafe.Sizeof(uintptr(0)) == 8 {
+		return uint64(runtime_memhash(ptr, uintptr(seed), uintptr(size)))
+	}
+	lo := runtime_memhash(ptr, uintptr(seed), uintptr(size))
+	hi := runtime_memhash(ptr, uintptr(seed>>32), uintptr(size))
+	return uint64(hi)<<32 | uint64(lo)
 }
 
-func AesHash(str string) uint64 {
-	return uint64(aeshashstr(unsafe.Pointer(&str), 0))
+//go:linkname runtime_memhash runtime.memhash
+//go:noescape
+func runtime_memhash(p unsafe.Pointer, seed, s uintptr) uintptr
+
+func StrHash(str string) uint64 {
+	hdr := (*reflect.StringHeader)(unsafe.Pointer(&str))
+	ptr := unsafe.Pointer(hdr.Data)
+	return uint64(rthash(ptr, len(str), shSeed))
 }
 
 func ByteHash(data []byte) uint64 {
-	if useAeshash {
-		return uint64(aeshashstr(unsafe.Pointer(&data), 0))
-	}
-	return uint64(strhash(unsafe.Pointer(&data), 0))
-}
-
-func AesByteHash(data []byte) uint64 {
-	return uint64(aeshashstr(unsafe.Pointer(&data), 0))
+	return uint64(rthash(unsafe.Pointer(&data[0]), len(data), shSeed))
 }

--- a/quickhash/strhash_test.go
+++ b/quickhash/strhash_test.go
@@ -40,8 +40,8 @@ func TestStrVSByte(t *testing.T) {
 	}
 
 	for _, k := range keys {
-		h1 := AesHash(k)
-		h2 := AesByteHash([]byte(k))
+		h1 := StrHash(k)
+		h2 := ByteHash([]byte(k))
 		if h1 != h2 {
 			t.Errorf("Expected str & byte hashes to be equal got: %d vs %d", h1, h2)
 		}
@@ -58,7 +58,7 @@ func BenchmarkStrHash(b *testing.B) {
 func BenchmarkAesHash(b *testing.B) {
 	x0 := "abcdefghijklmnopqrstuvwxyz"
 	for n := 0; n < b.N; n++ {
-		AesHash(x0)
+		StrHash(x0)
 	}
 }
 

